src/index.ts

```diff
@@ -486,18 +448,104 @@ export async function install(
       );
     }
   }
-  fs.writeFileSync(
-    path.join(destLoc, 'import-map.json'),
-    JSON.stringify({imports: importMap}, undefined, 2),
-    {encoding: 'utf8'},
-  );
+  await writeLockfile(path.join(destLoc, 'import-map.json'), importMap);
+  if (lockfile) {
+    await writeLockfile(path.join(destLoc, 'import-map.remote.json'), lockfile);
+  }
   Object.entries(assetObject).forEach(([assetName, assetLoc]) => {
     mkdirp.sync(path.dirname(`${destLoc}/${assetName}`));
     fs.copyFileSync(assetLoc, `${destLoc}/${assetName}`);
   });
   return true;
 }
 
+export async function installRemote(
+  installTargets: InstallTarget[],
+  lockfile: ImportMap | null,
+  pkgManifest: any,
+  config: SnowpackConfig,
+) {
+  const {dependencies} = config;
+  const downloadQueue = new PQueue({concurrency: 32});
+  const newLockfile: ImportMap = {imports: {}};
+
+  async function handleTopLevel(installSpecifier: string) {
+    // Right now, the CDN is only for JS packages. The CDN doesn't support CSS,
+    // non-JS assets, and has limited support for deep package imports. Snowpack
+    // will automatically fall-back failed/not-found assets from local
+    // node_modules/ instead.
+    const isPackage = !validatePackageName(installSpecifier).validForNewPackages;
+    if (isPackage && path.extname(installSpecifier)) {
+      return;
+    }
+    let installUrl: string;
+    if (lockfile && lockfile.imports[installSpecifier]) {
+      installUrl = lockfile.imports[installSpecifier];
+    } else {
+      const packageSemver: string | undefined =
+        (pkgManifest.peerDependencies || {})[installSpecifier] ||
+        (pkgManifest.devDependencies || {})[installSpecifier] ||
+        (pkgManifest.dependencies || {})[installSpecifier] ||
+        (dependencies || {})[installSpecifier];
+      if (!packageSemver) {
+        // TODO: warn
+        return;
+      }
+      if (packageSemver.includes(' ')) {
+        // TODO: warn
+        return;
+      }
+      if (
+        packageSemver.startsWith('npm:@reactesm') ||
+        packageSemver.startsWith('npm:@pika/react')
+      ) {
+        // TODO: warn
+        return;
+      }
+      installUrl = `${PIKA_CDN}/${installSpecifier}@${packageSemver}`;
```

Comment:
Tried building a local project I had with `--source pika` and it seemed to fail on `preact/hooks`. Here are the URLs it generated:

| Package | URL |
|:---------|:-----|
| `preact` | `https://cdn.pika.dev/-/preact@v10.3.2-BBkWjFhbfSgvwUixDhl2/dist=es2019/preact.js` |
| `preact/hooks` | `https://cdn.pika.dev/-/preact@vundefined/dist=es2019/preact.js` |

Reply:
Actually, scratch that. `preact` and `preact/hooks` work just fine. The problem is `preact-router`. I think because it is importing `preact` from within its package. For a remote `import` there that seems to yield the URL: `https://cdn.pika.dev/-/preact@vundefined/dist=es2019/preact.js`

Reply:
Looks like a bug in the CDN with preact, thanks for helping to find it! Fixed now 👍